### PR TITLE
Remove fallback for arm64e -> arm64.

### PR DESCRIPTION
**Pull request by varungandhi-apple**
Fixes rdar://problem/61407215

### CODE REVIEW

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

File: lib/Serialization/SerializedModuleLoader.cpp [hunk 1]
```diff
@@ -409,16 +405,20 @@ bool SerializedModuleLoader::maybeDiagnoseTargetMismatch(
...
     StringRef filePath = directoryIterator->path();
     StringRef extension = llvm::sys::path::extension(filePath);
-    if (file_types::lookupTypeForExtension(extension) ==
-          file_types::TY_SwiftModuleFile) {
+    auto fileType = file_types::lookupTypeForExtension(extension);
+    // We also check for interfaces here, because the SerializedModuleLoader
+    // is invoked after the ModuleInterfaceLoader; if the ModuleInterfaceLoader
+    // handled interfaces separately, we could get duplicate diagnostics.
+    if (fileType == file_types::TY_SwiftModuleFile
+        || fileType == file_types::TY_SwiftModuleInterfaceFile) {
```
Comment (brentdax) [inline lib/Serialization/SerializedModuleLoader.cpp:413]: Nice catch.

Nit: Should `||` be on the line before?

File: test/ModuleInterface/arm64e-fallback.swift [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// RUN: not %target-typecheck-verify-swift -target arm64e-apple-ios13.0 -F %S/Inputs | %FileCheck %s
+
+// REQUIRES: OS=ios
+
+import DummyFramework
+// CHECK: could not find module 'DummyFramework' for target 'arm64e-apple-ios'; found: arm64-apple-ios, arm64
```
Comment (brentdax) [inline test/ModuleInterface/arm64e-fallback.swift:6]: Can you use `// expected-error` here, or does it not work?

File: test/ModuleInterface/Inputs/DummyFramework.framework/Modules/DummyFramework.swiftmodule/arm64.swiftinterface [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// swift-interface-format-version: 1.0
+// swift-compiler-version: Apple Swift version 5.2
+// swift-module-flags: -target arm64-apple-ios13.0 -enable-library-evolution -swift-version 5 -enforce-exclusivity=checked -O -module-name DummyFramework
+import Foundation
+import Swift
+public struct S {
+  public var x: Swift.Int
+  public init()
+}
```
Comment (brentdax) [inline test/ModuleInterface/Inputs/DummyFramework.framework/Modules/DummyFramework.swiftmodule/arm64.swiftinterface:9]: What’s the purpose of all of these declarations? Are they part of the test?

File: test/ModuleInterface/arm64e-fallback.swift [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+public struct T {
```
Comment (brentdax) [inline test/ModuleInterface/arm64e-fallback.swift:8]: What’s the purpose of all of these declarations? Are they part of the test?

**Review by beccadax — CHANGES_REQUESTED** [review]
Might not actually need any changes—take a look.

File: test/ModuleInterface/arm64e-fallback.swift (hunk 2 above)
Comment (varungandhi-apple) [reply test/ModuleInterface/arm64e-fallback.swift:6]: There's a fatal error, so `%target-typecheck-verify-swift` fails (so you need a `not`) and `expected-error` doesn't do what you'd expect (e.g. even if the `expected-error` doesn't match, the `not` makes the test pass).

File: test/ModuleInterface/Inputs/DummyFramework.framework/Modules/DummyFramework.swiftmodule/arm64.swiftinterface (hunk 3 above)
Comment (varungandhi-apple) [reply test/ModuleInterface/Inputs/DummyFramework.framework/Modules/DummyFramework.swiftmodule/arm64.swiftinterface:9]: The point is to make the swiftinterface "realistic" (and not an edge case like being empty) which still being minimal. Same for the `.swift` file (e.g. the compiler could theoretically optimize the 0 declaration case to not actually import anything perhaps?). I'll remove it if you think it is unnecessary.

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 1 above)
Comment (varungandhi-apple) [reply lib/Serialization/SerializedModuleLoader.cpp:413]: For things we don't have written out (e.g. LLVM coding standards doesn't mention it) and don't seem to have a consistent style in the code (this file has both styles in use), I fallback (heh) to [PEP-8](https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator).

For example, here's some of the code you added recently in the same file:

```
    if (firstAbsoluteBaseName
        && maybeDiagnoseTargetMismatch(moduleID.Loc, moduleName,
                                       *firstAbsoluteBaseName)) {
```

Earlier, it was

```
    if (maybeDiagnoseTargetMismatch(moduleID.Loc, moduleName,
                                    primaryTargetSpecificName, currPath)) {
```

There's also examples of it being formatted with the operator at the end in the same file. 🤷

**Comment by varungandhi-apple** [comment]
> Might not actually need any changes

I don't get it. This branch shouldn't be present, should it?

```
  else if (Ctx.LangOpts.Target.getSubArch() ==
            llvm::Triple::SubArchType::AArch64SubArch_E) {
     body("arm64");
```

I'm removing it in the middle commit.

File: test/ModuleInterface/arm64e-fallback.swift (hunk 2 above)
Comment (harlanhaskins) [inline test/ModuleInterface/arm64e-fallback.swift:6]: This looks like it might also fix https://bugs.swift.org/browse/SR-10041

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

File: test/ModuleInterface/arm64e-fallback.swift (hunk 2 above)
Comment (varungandhi-apple) [reply test/ModuleInterface/arm64e-fallback.swift:6]: Thanks for pointing that out. 😄

File: test/ModuleInterface/Inputs/DummyFramework.framework/Modules/DummyFramework.swiftmodule/arm64.swiftinterface (hunk 3 above)
Comment (brentdax) [reply test/ModuleInterface/Inputs/DummyFramework.framework/Modules/DummyFramework.swiftmodule/arm64.swiftinterface:9]: I think they unnecessarily complicate the test, mainly by obscuring what it is about the compiler that you're testing.

If you want a minimum "realistic" swiftinterface, just have `import Swift` in it—every interface other than the standard library should have that.

**Review by beccadax — APPROVED** [review]

File: test/ModuleInterface/arm64e-fallback.swift (hunk 4 above)
Comment (varungandhi-apple) [reply test/ModuleInterface/arm64e-fallback.swift]: Done, removed.

**Comment by varungandhi-apple** [comment]
@swift-ci please smoke test

**Pull request merged**